File: problems/493-reverse-pairs/merge-sort.js
/**
 * 归并排序，每次数个数
 */
var reversePairs = function (nums) {
  nums = nums.slice()
  let res = 0

  // 归并排序
  function mergeSort (left, right) {
    if (right - left <= 1) return nums.slice(left, right)

    // 分两半
    const mid = left + (right - left >> 1)
    const A = mergeSort(left, mid)
    const B = mergeSort(mid, right)

    // 数个数
    let r = mid
    for (let l = left; l < mid; ++l) {
      while (r < right && A[l - left] > 2 * B[r - mid]) ++r
      res += r - mid
    }

    // 合并有序数组
    const C = []
    let a = 0, b = 0
    while (a < A.length && b < B.length) {
      if (A[a] < B[b]) {
        C.push(A[a++])
      } else {
        C.push(B[b++])
      }
    }
    if (a < A.length) C.push(...A.slice(a))
    if (b < B.length) C.push(...B.slice(b))

    return C
  }

  mergeSort(0, nums.length)
  return res
};

module.exports = reversePairs

File: problems/775-global-and-local-inversions/merge-sort.cpp
/**
 * 逆序对，使用归并排序
 * 
 * 时间：`O(NlogN)`, 460ms
 */

class Solution {
public:
    bool isIdealPermutation(vector<int>& arr) {
        int localCount = 0;
        for (int i = 1; i < arr.size(); ++i) {
            if (arr[i - 1] > arr[i]) {
                ++localCount;
            }
        }

        globalCount = 0;
        mergeSort(arr, 0, arr.size());

        return localCount == globalCount;
    }

    void mergeSort(vector<int> &arr, int left, int right) {
        if (right - left <= 1) return;

        int mid = left + (right - left >> 1);
        mergeSort(arr, left, mid);
        mergeSort(arr, mid, right);

        // 数逆序对
        int i = left, j = mid;
        for (; i < mid; ++i) {
            while (j < right && arr[j] <= arr[i]) ++j;
            globalCount += j - mid;
        }

        vector<int> tmpArr;
        i = left, j = mid;
        while (i < mid && j < right) {
            if (arr[i] < arr[j]) {
                tmpArr.push_back(arr[i++]);
            } else {
                tmpArr.push_back(arr[j++]);
            }
        }
        while (i < mid) {
            tmpArr.push_back(arr[i++]);
        }
        while (j < right) {
            tmpArr.push_back(arr[j++]);
        }

        // 写入arr
        int k = 0;
        for (int i = left; i < right; ++i) {
            arr[i] = tmpArr[k++];
        }
    }

private:
    int globalCount;
};